File: .agents/managed/java/src/main/java/com/example/corporate/security/StructuredLogger.java
package com.example.corporate.security;

import java.time.Instant;
import java.util.Locale;
import java.util.Map;
import java.util.Set;

public final class StructuredLogger {
    private static final Set<String> REDACTED = Set.of("key", "password", "secret", "token", "cookie", "authorization", "plaintext", "envelope");
    private StructuredLogger() {}

    public static void log(String level, String event, String message, Map<String, ?> context) {
        var builder = new StringBuilder(256).append('{')
            .append("\"timestamp\":\"").append(Instant.now()).append("\",")
            .append("\"level\":\"").append(escape(level.toUpperCase(Locale.ROOT))).append("\",")
            .append("\"event\":\"").append(escape(event)).append("\",")
            .append("\"message\":\"").append(escape(message)).append('"');
        for (var entry : context.entrySet()) {
            builder.append(",\"").append(escape(entry.getKey())).append("\":");
            var value = REDACTED.contains(entry.getKey().toLowerCase(Locale.ROOT)) ? "[REDACTED]" : entry.getValue();
            if (value instanceof Number || value instanceof Boolean) builder.append(value);
            else builder.append('"').append(escape(String.valueOf(value))).append('"');
        }
        System.out.println(builder.append('}'));
    }

    private static String escape(String value) {
        return value.replace("\\", "\\\\").replace("\"", "\\\"").replace("\n", "\\n").replace("\r", "\\r").replace("\t", "\\t");
    }
}

File: .agents/managed/kotlin/src/main/kotlin/com/example/corporate/security/StructuredLogger.kt
package com.example.corporate.security

import java.time.Instant

object StructuredLogger {
    private val redacted = setOf("key", "password", "secret", "token", "cookie", "authorization", "plaintext", "envelope")

    fun log(level: String, event: String, message: String, context: Map<String, Any?> = emptyMap()) {
        val fields = linkedMapOf<String, Any?>("timestamp" to Instant.now().toString(), "level" to level.uppercase(), "event" to event, "message" to message)
        context.forEach { (name, value) -> fields[name] = if (name.lowercase() in redacted) "[REDACTED]" else value }
        println(fields.entries.joinToString(prefix = "{", postfix = "}") { (name, value) -> "\"${escape(name)}\":${encode(value)}" })
    }

    private fun encode(value: Any?): String = when (value) {
        null -> "null"
        is Number, is Boolean -> value.toString()
        else -> "\"${escape(value.toString())}\""
    }

    private fun escape(value: String): String = value.replace("\\", "\\\\").replace("\"", "\\\"").replace("\n", "\\n").replace("\r", "\\r").replace("\t", "\\t")
}
